alphagov/di-account-management-backend
# OLH-1049 Send an audit event to TxMA when a user submits a report

## Proposed changes
OLH-1049 Send an audit event to TxMA when a user submits a report
 We create a Lambda that’s subscribed to the suspicious activity SNS topic, so that when the lambda receives a message, it will generate the audit event and send it to the audit event SQS queue. TxMA will pick up messages from the Audit Event Queue

### What changed

1. A new lambda that sends events to TxMA - This handler for this Lambda is built to be generic such that when another Lambda needs to send a message to TxMA(i.e. put a message on the audit event SQS queue, tall they will need to do is add a validation rule, add their transformation logic to convert the received event to a TxMA event and configure their Lambda within the infrastructure code, the transformation logic and validation rule is selected by an event name, so as part of the Lambda, the new event name would need to be configured as an environment variable.
2. A generic Validator function that will validate an event based on rules.
3. Error handling to ensure failures messages are sent to the associated DLQ


### Why did it change
So that reported suspicious activities can sent to TxMA

### Related links
https://govukverify.atlassian.net/browse/OLH-1049
https://govukverify.atlassian.net/browse/OLH-857
https://govukverify.atlassian.net/browse/OLH-1242
https://github.com/alphagov/di-account-management-backend/blob/main/docs/adr/0008-report-a-suspicious-activity.md


## Checklists
- The lambda should be able to subscribe to the SNS topic.
- The lambda should validate the fields .
- If validation fails send the message to the DLQ.
- If validation is a success , the lambda should then send the event to the audit event queue



### Environment variables or secrets
- EVENT_NAME - This must always be configured against any new lambda that uses the handler.
- [ ] No secrets were added or changed

### Permissions

## Testing
- Deployed to Dev
- Publish a suspicious activity message to the SNS topic
- Verify a TxMA valid event was sent to the  audit event SQS queue
- Verify a failure e.g. due to validation error as sent to the SendSuspiciousActivityToTxMADeadLetterQueue

## How to review
